### PR TITLE
[Feature] Batch-invariant support for speculative decoding

### PR DESCRIPTION
**Pull request by andakai**
## 1. Summary

This PR extends batch invariance to stochastic speculative decoding. A seeded
request now follows the same sampling and speculative-decoding path when batch
size, request order, or scheduling changes, including KV-cache preemption.

The implementation has two main parts:

~~1. Independent deterministic RNG streams for proposal sampling, acceptance,
   and residual sampling.~~
**Update on 8.27:** RNG-stream separation is handled separately in woosuk/mrv2-draft-gumbel-decouple branch. In particular, [commit 214a205](https://github.com/vllm-project/vllm/tree/woosuk/mrv2-draft-gumbel-decouple) separates draft Gumbel noise from acceptance/recovery noise. Since that issue is orthogonal to speculative batch invariance, I’ve removed the RNG-domain changes from this PR.

2. Preemption recovery that reconstructs the discarded proposal at the same
   logical boundary before verification.

The validated methods are EAGLE3, DFlash, and DSpark with probabilistic draft
sampling and standard rejection sampling. I didn't find a proper model with MTP to
test, because GDN hasn't been supported with batch-invariance now.

An exact open-PR search for speculative batch invariance found no duplicate.
Related issue: https://github.com/vllm-project/vllm/issues/27433. Broader searches 
found adjacent batch-invariance and speculative-decoding work,
but none covering this proposal RNG and preemption-recovery design. No issue
number was provided for an issue-specific duplicate check. AI assistance was
used for implementation, review, testing, and documentation. The human
submitter remains responsible for reviewing every changed line and reproducing
the reported tests before submission.

## 2. Method

~~### 2.1 Independent deterministic RNG streams~~

~~Proposal, acceptance, and residual sampling use separate RNG domains. Each draw
is derived from the request seed and its logical token position, so unrelated
requests cannot change its random sequence through batching or scheduling. The
bonus target token continues to use the normal target-sampling stream.~~

### 2.2 Preemption recovery and KV-cache boundary

A speculative proposal spans two scheduler steps: one step creates `K` draft
tokens, and the next verifies those tokens plus one bonus target token. If the
request is preempted between these steps, retaining only the request tokens is
not enough: the proposal and the worker-side hidden state used to create it have
been released.

For example, let the request contain `L = 4` committed tokens:

```text
A B C D
```

Before preemption, the proposal is produced as follows:

```text
execute C -> target hidden state h_C
sample D from h_C
draft from (h_C, D) -> P1 P2
```

The logical state is therefore:

```text
computed tokens:       A B C
sampled next token:          D
pending proposal:              P1 P2
```

On preemption, the scheduler records only the proposal length (`K = 2`) and
uses the normal cleanup path to release KV-cache blocks and worker state. On
recovery it performs:

```text
prefix-cache hit:      A B       # at most L - 2 tokens
replay:                    C     # reconstruct h_C
known next token:             D
regenerate:          (h_C, D) -> P1 P2
verify atomically:               P1 P2 + bonus token
```

An `L - 1` hit would cache `A B C` and skip the forward pass for `C`. Its KV
entries are available, but `h_C`, which the drafter needs, is not. Executing
`D` instead would advance to `h_D` and build the proposal for the next logical
position. Limiting the hit to `L - 2` therefore recreates the same `(h_C, D)`
input and, with the position-based draft RNG stream, the same proposal as the
non-preempted execution. Scheduling the subsequent `K + 1` verification as one
unit prevents another partial boundary from being introduced.

### 3.1 Serving configuration

The two runs for each method use the same target/speculative model pair and
change only `VLLM_BATCH_INVARIANT`.

| Method | Target model | Speculative model | Speculative tokens |
| --- | --- | --- | ---: |
| EAGLE3 | `Qwen/Qwen3-8B` | `RedHatAI/Qwen3-8B-speculator.eagle3` | 3 |
| DFlash | `Qwen/Qwen3-8B` | `z-lab/Qwen3-8B-DFlash-b16` | 16 |
| DSpark | `Qwen/Qwen3-4B-FP8` | `deepseek-ai/dspark_qwen3_4b_block7` | 7 |

The command below uses placeholders so that the same serving setup can be
reused for each row.

```bash
export TARGET_MODEL="TARGET_MODEL_ID"
export DRAFT_MODEL="DRAFT_MODEL_ID"
export SPEC_METHOD="eagle3"       # eagle3, dflash, or dspark
export SPEC_TOKENS="3"            # 3, 16, or 7 respectively
export SPEC_CONFIG="{
  \"model\": \"${DRAFT_MODEL}\",
  \"method\": \"${SPEC_METHOD}\",
  \"num_speculative_tokens\": ${SPEC_TOKENS},
  \"draft_sample_method\": \"probabilistic\",
  \"rejection_sample_method\": \"standard\",
  \"enable_adaptive_verification\": false
}"

# BI off
VLLM_USE_V2_MODEL_RUNNER=1 VLLM_BATCH_INVARIANT=0 \
vllm serve "$TARGET_MODEL" \
  --max-model-len 2048 \
  --enable-prefix-caching \
  --enable-chunked-prefill \
  --async-scheduling \
  --speculative-config "$SPEC_CONFIG"

# BI on
VLLM_USE_V2_MODEL_RUNNER=1 VLLM_BATCH_INVARIANT=1 \
vllm serve "$TARGET_MODEL" \
  --max-model-len 2048 \
  --enable-prefix-caching \
  --enable-chunked-prefill \
  --async-scheduling \
  --speculative-config "$SPEC_CONFIG"
```

For DSpark, the evaluation added `"attention_backend":"FLASH_ATTN"` to
`SPEC_CONFIG`. DFlash used `--max-num-seqs 128` for model warmup.

### 3.2 GSM8K end-to-end results

All 1,319 GSM8K test examples were served twice per configuration: concurrency
32 in natural order and concurrency 64 in reverse order. Every request used a
stable unique seed with temperature `0.6`, top-k `20`, top-p `0.95`, and at most
256 output tokens. Exact BI requires identical token IDs, text, and finish
reason between the two phases. Accuracy and throughput below are from the
concurrency-64 phase.

| Method | BI | Exact BI | Mismatches | Accuracy | Req/s | Output tok/s |
| --- | --- | --- | ---: | ---: | ---: | ---: |
| EAGLE3 | off | FAIL | 401 | 88.32% | 28.92 | 3468.76 |
| EAGLE3 | on | PASS | 0 | 88.02% | 21.68 | 2614.11 |
| DFlash | off | FAIL | 187 | 87.79% | 11.08 | 1335.24 |
| DFlash | on | PASS | 0 | 87.79% | 10.50 | 1270.43 |
| DSpark | off | FAIL | 118 | 83.17% | 42.67 | 5398.54 |
| DSpark | on | PASS | 0 | 82.94% | 41.28 | 5169.72 |

| Method | Accuracy delta | Req/s delta | Output tok/s delta |
| --- | ---: | ---: | ---: |
| EAGLE3 | -0.30 pp | -25.03% | -24.64% |
| DFlash | +0.00 pp | -5.22% | -4.85% |
| DSpark | -0.23 pp | -3.26% | -4.24% |

These throughput runs validate invariance across concurrency and request order;
they did not trigger KV-cache preemption. The focused E2E test below uses a
small KV-cache allocation and asserts that preemption occurred before comparing
serial and batched token IDs.

### 3.3 Minimal tests

The first test checks independent RNG domains, batch splitting, and the target
bonus-token stream. The second forces speculative decoding and KV-cache
preemption, then checks exact batch invariance.

```bash
#!/usr/bin/env bash
set -euo pipefail

.venv/bin/python -m pytest -q \
  tests/v1/spec_decode/test_rejection_sampler_utils.py::test_spec_rng_streams_are_batch_invariant \
  tests/v1/e2e/spec_decode/test_stochastic_batch_invariance.py::test_spec_batch_invariance_with_preemption
```

The E2E test requires one CUDA GPU with at least 32 GB of memory and access to
the configured public model artifacts.

### CODE REVIEW

File: vllm/config/vllm.py [hunk 1]
```diff
@@ -1043,6 +1043,22 @@ def __post_init__(self):
...
+                not self.use_v2_model_runner
+                or not speculative_config.supports_batch_invariance()
+            )
+        ):
+            raise ValueError(
+                "VLLM_BATCH_INVARIANT only supports speculative decoding with "
+                "Model Runner V2, EAGLE3/DFlash/DSpark, probabilistic drafting, "
+                "standard rejection sampling, fixed speculative lengths, and "
+                "adaptive verification disabled."
+            )
```
Comment (yewentao256) [inline vllm/config/vllm.py]: https://github.com/vllm-project/vllm/blob/a62aeef54c7060bb99460d6ae48d3abca6d93bf0/tests/v1/distributed/test_eagle_dp.py#L43-L72

This test hard code the eagle?

File: vllm/v1/core/sched/scheduler.py [hunk 2]
```diff
@@ -473,6 +490,22 @@ def _reserve_prefill_lookahead(
...
+        num_new_tokens: int,
+    ) -> int:
+        if request.request_id not in self.spec_recovery_sizes:
+            return num_new_tokens
+        if not request.spec_token_ids:
+            return min(
+                num_new_tokens,
+                request.num_tokens - 1 - num_computed_tokens,
+            )
+        required = len(request.spec_token_ids) + 1
```
Comment (yewentao256) [inline vllm/v1/core/sched/scheduler.py]: Please take a look here, will required only equal to k+1 in all cases?

**Review by yewentao256 — COMMENTED** [review]
Thanks for the work! Also CC @benchislett for the spec decode part

**Comment by yuvalluria** [comment]
Preemption recovery approach is clean. One gap for the record: MTP/GDN (hybrid linear-attn/MoE) is not covered — the Mamba scan's h_C is also M-sensitive on recovery. Will test against that once PR #45819 merges and report back.

Worth updating the PR description to reflect that RNG-stream separation is out of scope here — the current text still implies broader RNG coverage than what's actually in this PR.

Question: is the -25% req/s on EAGLE3 from the preemption recovery path itself, or only when batch_invariant=1 is set?

Signed-off-by: Yuval Luria <yluria@redhat.com>

**Comment by WenyueD** [comment]
Additional composition case for this work: Qwen3.8-27B + DFlash2 on A800.
The batch-invariant test is currently blocked by the GDN_ATTN rejection covered
by #42960 / #45819. Evidence and exact-tie analysis:
https://github.com/vllm-project/vllm/issues/42960#issuecomment-5491990078

Once GDN support is available, this model pair would be useful additional
coverage for speculative batch invariance.

**Comment by andakai** [comment]
Thanks for all the advice and discussion! I’m currently refactoring the scheduler, as it has accumulated quite a lot unnecessary codes. I’m trying to simplify it while also addressing the comments today.

As for GDN, I believe this approach should work for it as well, since the design is fairly general. I will also test it when the GDN BI pr gets merged.

**Comment by andakai** [comment]
I found that, in my original gsm8k evaluation, `num_preemptions_total` was zero, so recovery never worked. 0 mismatch mainly came from forcing batch-invariant requests to use the same Triton top-k/top-p implementation.

```diff
-if HAS_TRITON and logits.shape[0] >= 8:
+if HAS_TRITON and (envs.VLLM_BATCH_INVARIANT or logits.shape[0] >= 8):
```

To test the effect of recovery, I limit the kvcache size to make preempt happen. each configuration served the same 1,319 prompts twice: first at concurrency 32 in natural order, then at concurrency 64 in reverse order. Both phases used `max_num_seqs=128`, `max_num_batched_tokens=8192`, `max_model_len=2048`, and 768 KV-cache blocks.

| Implementation | Method | BI | Mismatched prompts | Preemptions (concurrency 32) | Preemptions (concurrency 64) | Concurrency-64 elapsed |
| --- | --- | --- | ---: | ---: | ---: | ---: |
| current PR | EAGLE3 | off | 419 | 0 | 34 | 54.2s |
| current PR | EAGLE3 | on | **0** | 0 | 38 | 65.5s |
| current PR | DFlash | off | 373 | 0 | 394 | 129.2s |
| current PR | DFlash | on | **0** | 0 | 383 | 140.4s |
| current PR | DSpark | off | 619 | 0 | 220 | 34.3s |
| current PR | DSpark | on | **0** | 0 | 215 | 35.1s |
| Triton Sampler-only, no recovery | DSpark | off | 638 | 0 | 263 | 34.4s |
| Triton Sampler-only, no recovery | DSpark | on | **39** | 0 | 186 | 34.8s |

File: vllm/config/vllm.py (hunk 1 above)
Comment (andakai) [reply vllm/config/vllm.py]: I removed the method hardcoding now.  It is now method agnostic. Actually not tested on mtp yet, but I think it can work the same. I can test it later when the related pr gets merged.

File: vllm/v1/core/sched/scheduler.py (hunk 2 above)
Comment (andakai) [reply vllm/v1/core/sched/scheduler.py]: The code has been changed, but the logis hasn't. Two questions here from my understanding? why "k+1" and whether "all cases"?

For normal verification, it is k + 1: one committed token plus k drafts. If it does not fit, we defer it. truncation could make the same position a draft token in one batch but a bonus token in another. This can definitely cause different tokens.

With async scheduling, the required count is k + num_output_placeholders + num_tokens - num_computed_tokens, so it is not always k + 1.
